mongo/description/server_selector.go

```diff
@@ -161,6 +161,40 @@ func ReadPrefSelector(rp *readpref.ReadPref) ServerSelector {
 	})
 }
 
+// OutputSelector selects servers based on the provided read preference given that the underlying operation is
+// aggregate with an output stage.
+func OutputSelector(rp *readpref.ReadPref) ServerSelector {
```

Comment:
`OutputSelector` is a copy of `ReadPrefSelector` except it calls `selectForRSWithOutput` instead of `selectForReplicaSet`.

Reply:
Because `OutputSelector` is almost an exact copy of `ReadPrefSelector`, I think it's better to factor out the common code into a shared function.

E.g. possible function (including recommended changes to `selectForReplicaSet`):
```go
func readPrefSelector(rp *readpref.ReadPref, isOutputStage bool) ServerSelector {
	return ServerSelectorFunc(func(t Topology, candidates []Server) ([]Server, error) {
		// ...
		case ReplicaSetNoPrimary, ReplicaSetWithPrimary:
			return selectForReplicaSet(rp, t, isOutputStage, candidates)
		// ...
	})
}
```

Then, `ReadPrefSelector` and `OutputSelector` become:
```go
func ReadPrefSelector(rp *readpref.ReadPref) ServerSelector {
	return readPrefSelector(rp, false)
}

func OutputSelector(rp *readpref.ReadPref) ServerSelector {
	return readPrefSelector(rp, true)
}
```

Reply:
Ah excellent idea. Done.